---
output:
  pdf_document:
    highlight: tango
    number_sections: true
    fig_crop: no
fontsize: 11pt
---

```{r, echo=FALSE}
# initialisé R 
rm(list=ls())
# options pour le document word et les images générées
knitr::opts_chunk$set(echo=TRUE, warning=FALSE, message=FALSE, fig.align="center", dev="png", fig.path="Images/")
# répertoire de travail
source("/<write_correct_path_here>/AskoR.R")
```
\begin{center}
\textbf{{\huge Pipeline askoR}}
\end{center}
\begin{center}
\textbf{{\huge Differential Expression Analysis Tool}}
\end{center}
\begin{center}
\textbf{Susete Alves C., Kevin Gazengel and Fabrice Legeai}
\end{center}
\begin{center}
{\textit{\small INRAE Centre Bretagne - Normandie, UMR 1349 iGEPP}}
\end{center}

\vspace{24pt}

# ABSTRACT
**AskoR** is a pipeline for the analysis of gene expression data, using edgeR.
Several steps are performed: data filters (cpm method), normalize this filtered data, look at the correlation of our data, run differential expression analysis, compare contrast, GO enrichment and co-expression.

# QUICK START
Here we show the most basic steps for a differential expression analysis. The different parameters and input/output files will be described in detail later. First of all, you must have in your workspace a folder **"input "** _(case sensitive)_. 
This folder should contain at least the following files:

  - tabulated text files, **one file per sample**, containing the counts of reads for each gene,

 Geneid | Counts
  :-|:-|
 gene1 | 23   |
 gene2 | 355  |
 gene3 | 0    |
 ...   |
 geneN | 3605 | 

  - a tabulated text file describing the design of experiments:

 sample | condition | time | genotype | ... | file |
 :-|:-|:-|:-|:-|:-|
 ech1 | cond1 | T0 | Wt1 | ... | ech1_counts.txt |
 ech2 | cond2 | T0 | Wt2 | ... | ech2_counts.txt |
 ech3 | cond3 | T2 | Mt1 | ... | ech3_counts.txt |
 ... |
 echN | condN | T5 | Mt4 | ... | echN_counts.txt |

\newpage

  - a tabulated text file describing the contrasts wanted: 

 Condition | condi1vscondi2 | condi1vscondi3 | Condi2vsCondi3 | ... |
 :-|:-:|:-:|:-:|:-:|
 condi1 | + | + | 0 | ... |
 condi2 | - | 0 | + | ... |
 condi3 | 0 | - | - | ... |
 ... |

In a second step, you will have to source your working directory and then initialize the parameters&nbsp;:
```{r set_parameters, eval=FALSE}
setwd("/work_directory/")
parameters<-Asko_start() 
```
The object **"parameters "** will contain all the useful parameters for the pipeline. Now enter the names of the files located in the folder **"input "**:
```{r input_files, eval=FALSE}
# matrix of different contrasts desired
parameters$contrast_file = "Contrasts.txt"  
# file describing samples / experience design
parameters$sample_file = "Samples.txt"      
```
You do not need to specify the folder name (i.e. _input/Contrats.txt_) since, by default, it will search for it.
You don't need to specify our count files since they are specified in the **"Samples.txt "** file. Once this is done, you can run the scan as follows:
```{r quick_start, eval=FALSE}
# load all data and parameters
data<-loadData(parameters) 
# check sample description and contrast
asko_data<-asko3c(data, parameters)
# describe datasets and filters out under-expressed genes
asko_filt<-GEfilt(data, parameters)
# normalize filtered datasets
asko_norm<-GEnorm(asko_filt, asko_data, data, parameters)
# check correlation between datasets
GEcorr(asko_norm,parameters)
# run analysis of gene expression using edgeR
resDEG<-DEanalysis(asko_norm, data, asko_data, parameters)
```
\newpage

# USER GUIDE
You'll find a test set in the **example** folder. It'll be used for the rest of the documentation.\
This folder contains a **run_asko.R** file that I will describe here, you will find the **input** folder with the input files we need for the study: 

  - Count matrix file: CountsMatrix.txt **OR** Counts files per samples: files in "counts" directory
  - Samples file: Samples_CountsMatrix.txt **OR** Samples_CountsFiles.txt 
  - Contrasts file: Contrasts.txt
  - Genes annotations file: Genes_annotations.txt (optional)
  - GO annotations file: GO_annotations.txt (optional)

**IMPORTANT :** All input files must be in a folder named **input** _(case sensitive)_.

## Input files description
### Count files
#### Sample count files
You have a count file **per sample**, they can be in text or csv format. In this case, you will have to fill in the **Samples** file with the path and name of the **count files** for each sample in a _**"file "**_ column. This one can contain several columns according to the counting tools used, you will have to inform the following parameters:

  - ```parameters$col_genes``` &nbsp;&nbsp; column number with GeneId (default 1)
  - ```parameters$col_counts``` &nbsp;&nbsp; column number with count (default 7)
  - ```parameters$sep``` &nbsp;&nbsp; column separator (default **"\\t"** ) 

Example of count file:

 Geneid | Chr | Start | End | Strand | Length | Counts
 :-|:-|-:|-:|:-:|-:|-:
 Gene_000001 | Random_Chr_001 | 1692 | 1907 | - | 215 | 0
 Gene_000002 | Random_Chr_001 | 6641 | 8705 | - | 2064 | 43.5
 Gene_000003 | Random_Chr_001 | 9228 | 9569 | - | 341 | 8
 Gene_000004 | Random_Chr_001 | 12009 | 13155 | - | 1146 | 781
 Gene_000005 | Random_Chr_001 | 15242 | 15844 | + | 602 | 16
 Gene_000006 | Random_Chr_001 | 16304 | 19834 | + | 3530 | 9
 Gene_000007 | Random_Chr_001 | 20595 | 21625 | - | 1030 | 13.83
 Gene_000008 | Random_Chr_001 | 22377 | 23461 | - | 1084 | 565.33
 **...** |

The corresponding parameters:
```{r counts_files, eval=FALSE}
parameters$col_genes=1
parameters$col_counts=7
parameters$sep="\t"
```
In this example column 7 contains the counts and the gene identifiers are in the first column, it's a tabulate file so the column separator is <TAB> encoded "\\t".

#### Counts matrix file
It is also possible to have a table, **tabulated file**, grouping the counts for each gene in each sample, in text or csv format. The **Samples** file should not contain a **file** column, you will have to fill in the name of the count file: ```parameters$fileofcount```.\
\
Example of a count matrix, the column separator is a tabulation:

 Geneid | AC1R1 | AC1R2 | AC1R3 | BC1R1 | BC1R2 | BC1R3 | **...**
 :-|-:|-:|-:|-:|-:|-:|:-:
 Gene_000001 | 0 | 1 | 0 | 0 | 0 | 1 | **...**
 Gene_000002 | 43.5 | 25.33 | 31.5 | 27.5 | 29.5 | 29 | **...**
 Gene_000003 | 8 | 4 | 5 | 30 | 16 | 13 | **...**
 Gene_000004 | 781 | 412 | 626 | 558 | 538 | 346 | **...**
 Gene_000005 | 16 | 7 | 13 | 9 | 8 | 6 | **...**
 Gene_000006 | 9 | 4 | 5 | 21 | 15 | 12 | **...**
 **...** |


### Samples file
This **tabulated file** describes the design of experiments. The first and second columns are mandatory and are named **"sample "** and **"condition "**. You may have several other columns. The contents of the condition column will be the same as in the **Contrast**file.\ 
\
The column **"color "** is optional, it allows to predefine the color of the sample in the graphs. If it is absent askoR will assign colors itself.
\
The column **"file "** is mandatory if you have samples counts files. In the example below, these files are grouped in a _"counts"_ folder. You do not need to specify the name of the **"input "** folder (i.e. input/counts/AC1R1_counts.txt) since, by default, it will search for it.
\
Don't forget to fill in the name of your Samples file: ```parameters$sample_file```, no need to specify the "input" folder.\
\
Eample of a Samples.txt file:

 sample | condition | genotype | treatment | color | file
 :-|:-|:-|:-|:-|:-
 AC1R1 | AC1 | A | C1 | darkorchid2 | counts/AC1R1_counts.txt
 AC1R2 | AC1 | A | C1 | darkorchid2 | counts/AC1R2_counts.txt
 AC1R3 | AC1 | A | C1 | darkorchid2 | counts/AC1R3_counts.txt
 BC1R1 | BC1 | B | C1 | saddlebrown | counts/BC1R1_counts.txt
 BC1R2 | BC1 | B | C1 | saddlebrown | counts/BC1R2_counts.txt
 **...** |


### Contrast file
This **tabulated file** indicates contrasts you wish to make between your different **conditions**.
The first column corresponds to the *condition* column of the **Samples** file, then the others are columns the comparisons to be made in the form **ConditionXvsConditionY**. Then under these columns, **ConditionX** will be noted **+** and **ConditionY** will be noted **-**, the rest **0**. You will have to fill in the name of your file: ```parameters$contrast_file```\
\
Example of contrasts file:

 Condition | AC1vsAC2 | AC1vsAC3 | AC2vsAC3 | BC1vsBC2 | **...**
 :-|:-|:-|:-|:-|:-
 AC1 | + | + | 0 | 0 | **...**
 AC2 | - | 0 | + | 0 | **...**
 AC3 | 0 | - | - | 0 | **...**
 BC1 | 0 | 0 | 0 | + | **...**
 BC2 | 0 | 0 | 0 | - | **...**
 **...** |
\

#### Genes annotations file
This **tabulated file** contains the annotations of your genes, it is optional. It can contain several columns but the first one must be the gene identifier.  You will have to fill in the name of your file: ```parameters$annotation```.\
\
Example of annotation file:

 SeqName | Description
 :-|:-
 Gene_000001 | hypothetical protein pbra 009537
 Gene_000002 | hypothetical protein pbra 009324
 Gene_000003 | histone-lysine n-methyltransferase nsd2
 Gene_000004 | hypothetical protein pbra 009496
 **...** |
\

### GO annotations file 
This **tabulated file** will be **WITHOUT HEADER**, the first column contains the gene identifier and the second column contains all the corresponding GOs separated by a **comma**. This file is optional, you will have to fill in its name: ```parameters$geneID2GO_file```. _Cf. GO enrichment Section._\
\
Example of GOs annotation file:

  | 
 :-|:-
 Gene_000001 | GO:0003676,GO:0015074
 Gene_000002 | GO:0003676,GO:0015074
 Gene_000003 | GO:0005488,GO:0006807,GO:0016740,GO:0043170,GO:0044238
 Gene_000005 | GO:0005525,GO:0005525,GO:0005525
 **...** |
\

## Out files
All the generated files and images will be in a folder named by default **"DE_analysis "**, you can change this name: ```parameters$analysis_name```.

## Script run_askoR.R
### Initialize and load data
Now that we have our input files, look at the script, you should have these first lines:
```{r run_askor1, eval=FALSE}
# Path to askoR file 
source("/home/user/test/AskoR/askoR.R")
# Workdir path
setwd("/home/user/Workspace/tests_askoR/")
# Sets defaults parameters
parameters<-Asko_start()
```
Don't forget to replace the paths, the first one to the _askoR.R_ script and the second one to your working directory (containing the **input** folder). \
\
Once this step has been completed, you will be able to indicate the names of the analysis files:
```{r run_askor2, eval=FALSE}
# output directory name (default DE_analysis)
parameters$analysis_name="DEG_test"

# input files:
# matrix of different contrasts desired
parameters$contrast_file = "Contrasts.txt"    
# file containing the functional annotations for each gene
parameters$annotation = "Genes_annotations.txt"      
# GO annotation files
parameters$geneID2GO_file = "GO_annotations.txt"   
```
 - If you use a counts matrix :
```{r run_askor3a, eval=FALSE}
# matrix of count for all samples/conditions
parameters$fileofcount = "CountsMatrix.txt"  
# file describing all samples
parameters$sample_file = "Samples_CountsMatrix.txt"  
```
 - If you use files of counts :
```{r run_askor3b, eval=FALSE}
# file describing all samples
parameters$sample_file = "Samples_CountsFiles.txt"
# column with the gene names (default 1)
parameters$col_genes = 1          
# column with the counts values (default 7)
parameters$col_counts = 7 
# field separator (default "\t")
parameters$sep = "\t" 
```
We are informed that two samples, "AC3R2" and "BC3R3", had problems during the experiments, it is requested to extract it from our analysis. No need to redo all the files, just use the parameter: ```parameters$rm_sample```.  You can provide a list of samples **"c("sample1","sample2","sample3",...)**, or a single sample **c("sample1")**. In the same way, if you only want to work on a part of your samples, you can use ```parameters$select_sample```.
```{r run_askor3c, eval=FALSE}
# delete sample AC3R2 
parameters$rm_sample = c("AC3R2","BC3R3")
```
It's time to load your data:
```{r run_askor4, eval=FALSE}
data<-loadData(parameters)
```
```{r exec1, echo=FALSE}
rm(list=ls())
source("/<write_correct_path_here>/AskoR.R")
setwd("/<write_correct_path_here>/JeuTest/")
parameters<-Asko_start()
parameters$analysis_name="DEG_test"
parameters$fileofcount = "CountsMatrix.txt"  
parameters$sample_file = "Samples_CountsMatrix.txt"  
parameters$contrast_file = "Contrasts.txt"    
parameters$annotation = "Genes_annotations.txt"  
parameters$geneID2GO_file = "GO_annotations.txt"
parameters$rm_sample = c("AC3R2","BC3R3")
data<-loadData(parameters)
``` 
You can see that two folders have been created in DEG_test:
  
 - **images:** You can see that two folders have been created in DEG_test.
 - **Askomics:** will contain the results of the differential analysis for each contract. All these files are in a format readable by askomics. 

Then the samples and conditions that have been loaded are displayed. These have been loaded into a structure called **"data "**.\
\newpage Some sample commands to display your data:
```{r run_askor4b}
# Displays all samples recorded
data$samples  

# Displays all contrast recorded
data$contrast 
```
\newpage
```{r run_askor4c}
# Displays design experiment
data$design   

# Displays the first 4 lines of counts table.
head(data$dge$counts,n=4) 

# Total number of genes:
dim(data$dge$counts)[1]

# Total number of samples:
dim(data$dge$counts)[2]

# summary of CPM by samples:
summary(cpm(data$dge))
```

The next step is to generate the files describing your experiences for Askomics. Even if you don't plan to use Askomics, this command is **mandatory** because it generates a data structure **"asko_data "** that will be used in the further analysis.
```{r run_askor5}
asko_data<-asko3c(data, parameters)
```

### Filtering data
For the filters the CPM method is used, you can set the cutoff values you want to:
```{r run_askor6a}
# CPM's threshold 
parameters$threshold_cpm = 0.5  
# minimum of sample which are upper to cpm threshold 
parameters$replicate_cpm = 3 # we have 3 replicates
```

```{r run_askor6b}
# run filtering 
asko_filt<-GEfilt(data, parameters)

# Total number of filtered genes: 
dim(asko_filt$counts)[1]
```
The filtered data is saved in a structure called here: **asko_filt**. In the folder **DEG_test/images/**, you should find the images representing your data before and after filtering.
```{r filters1a, echo=FALSE, out.width='48%', out.height='48%', fig.cap="Filtering Data", fig.align='center', fig.show='hold', fig.keep='all'}
par(mar=c(0,0,0,0), oma=c(0,0,0,0))
knitr::include_graphics(c("DEG_test/images/DEG_test_boxplot_logcpm_before_filtering.png","DEG_test/images/DEG_test_boxplot_logcpm_after_filtering.png","DEG_test/images/DEG_test_barplot_logcpm_before_filtering.png","DEG_test/images/DEG_test_barplot_logcpm_after_filtering.png"))
```
```{r filters1b, echo=FALSE, out.width='48%', out.height='30%', fig.cap="Density Graphes", fig.align='center', fig.show='hold', fig.keep='all'}
par(mar=c(0,0,0,0), oma=c(0,0,0,0))
knitr::include_graphics(c("DEG_test/images/DEG_test_raw_data_1.png","DEG_test/images/DEG_test_filtered_data_1.png"))
```
\newpage You notice that the legend of the density graphs is very low compared to the graph. You can correct this with the options ```parameters$densinset``` which modifies the position of the legend, it is also possible to define the number of columns with ```parameters$legendcol```. Finally, restart "GEfilt" function.
```{r param1}
# Set position the legend in bottom density graphe
parameters$densinset = 0.20
# Set numbers of column for legends
parameters$legendcol = 8
# run filtering 
asko_filt<-GEfilt(data, parameters)
```
```{r filters2, echo=FALSE, out.width="49%", out.height="49%", fig.cap="Density Graphes corrected", fig.show='hold', fig.align='center', out.extra='trim={0 7cm 0 0cm},clip'}
knitr::include_graphics(c("DEG_test/images/DEG_test_raw_data.png","DEG_test/images/DEG_test_filtered_data.png"))
```

### Normalize data
Once the filters have been made, we can proceed to the normalization of the data. At this step, you can generate file with normalize factor values for each sample ```parameters$norm_factor=TRUE``` and/or generate with normalize counts ```parameters$norm_counts=TRUE```. 
```{r run_askor7}
# run normalization
asko_norm<-GEnorm(asko_filt, asko_data, data, parameters)
```
Normalized data is saved in a structure called here : **asko_norm**. In the folder **DEG_test/images/**, you should find the images representing your data after normalization. Two files are automatically generated because they will be used for co-expression analysis: _"DEG_test_CPMNormCounts.txt"_ and _"DEG_test_CPM_NormMeanCounts.txt"_.
```{r norm, echo=FALSE, out.width="50%", out.height="49%", fig.cap="Normalization graphs (barplot)", fig.show='hold', fig.align='center'}
knitr::include_graphics("DEG_test/images/DEG_test_boxplot_logcpm_after_norm.png")
```
```{r norm2, echo=FALSE, out.width="49%", out.height="49%", fig.cap="Normalization graphs (heatmap)", fig.show='hold', fig.align='center'}
knitr::include_graphics(c("DEG_test/images/DEG_test_heatmap_CPMcounts_per_sample.png","DEG_test/images/DEG_test_heatmap_meanCounts_per_condi.png"))
```
\newpage

### Correlation
From the filtered and normalized data, we can re-correlate the correlation between our samples.
```{r run_askor8}
GEcorr(asko_norm,parameters)
```
Several graphics will be saved in the **DEG_test/images/** folder, including MDS and PCA plots. Axis1 vs axis2 differentiate our A and B samples.
```{r corr, echo=FALSE, out.width="49%", out.height="49%", fig.cap="MDS and PCA plots - axis 1 and 2", fig.show='hold', fig.align='center'}
knitr::include_graphics(c("DEG_test/images/DEG_test_MDS_corr_axe1_2.png","DEG_test/images/DEG_test_PCA_axe1_2.png"))
```
\newpage

### Differential expression analysis
The differential expression analysis can be started. We will play with the following parameters:

 - FDR threshold value 
 - logFC threshold value
 - normalization method (TMM/RLE/upperquartile/none)
 - p-value adjust method (holm/hochberg/hommel/bonferroni/BH/BY/fdr/none)
 - GLM method (lrt/qlf) 

```{r run_askor9a}
# FDR threshold
parameters$threshold_FDR = 0.05      
# logFC threshold
parameters$threshold_logFC = 0  
# normalization method 
parameters$normal_method = "TMM"     
# p-value adjust method
parameters$p_adj_method = "BH"      
# GLM method
parameters$glm = "lrt"
```

You can decide to get the Volcano or Mean-Difference Plots for each contrast: 
```{r run_askor9b}
# Mean-Difference Plot of Expression Data (aka MA plot) 
parameters$plotMD = T 
# Volcano plot for a specified coefficient/contrast of a linear model 
parameters$plotVO = T 
```

Once our parameters are defined, we can start the analysis.\
```{r run_askor9c}
# run differential expression analysis
resDEG<-DEanalysis(asko_norm, data, asko_data, parameters)
```
\
\
For each contrast, you will find the number of over- or under-expressed genes. Genotype B does not show any major effects of the treatment, unlike genotype A. We also observe a certain number of genes differentially expressed between the genotypes.\
\
A file named **"DEG_test_summary_DE "** is located in the **DEG_test/** folder, which contains for each gene whether it is over-expressed (**1**) or under-expressed (**-1**) or neutral (**0**) for a given contrast. If you had provided an annotation file, these will be found in the last columns.\
\
First lines of the :

 |  | AC1vsAC2 | AC1vsAC3 | **...** | AC2vsBC2 | AC3vsBC3 | Description
 :-|:-:|:-:|:-:|:-:|:-:|:-
 Gene_000002 | 0 | 0 | **...** | 0 | 0 | hypothetical protei...
 Gene_000003 | -1 | 0 | **...** | 0 | 0 | histone-lysine n-m...
 Gene_000004 | 1 | 1 | **...** | -1 | 0 | hypothetical prote...
 **...** |

You'll find in **"DEG_test/images/"** directory, les Volcano, MD plots and heatmap for each contrast.
```{r dge, echo=FALSE, out.width="49%", out.height="49%", fig.cap="DEanalysis plots", fig.show='hold', fig.align='center'}
knitr::include_graphics(c("DEG_test/images/AC3vsBC3_MeanDifference_of_ExpressionData.png","DEG_test/images/AC3vsBC3_VolcanoPlot.png","DEG_test/images/AC3vsBC3_topDGE_heatmap.png"))
```

\newpage
## Basic comparisons of lists of differentially expressed genes.
You can compare your lists of differentially expressed genes using two methods: Venn diagrams or Upset graphs. Venn diagrams allow you to compare up to 4 lists while Upset allows you to make wider comparisons. However, if you have too many lists to display the graph may be unreadable.

### Venn diagram
To display the Venn diagrams, you need to specify the type of comparison wanted ```parameters$VD```:

 - "all"  : Create VennDiagrams for all differentially expressed genes
 - "up"   : Create VennDiagrams for gene expressed UP
 - "down" : Create VennDiagrams for gene expressed DOWN
 - "both" : Create VennDiagrams for gene expressed UP and DOWN (in the same graph) 

Next, you must provide a list of the comparisons to display: ```parameters$compaVD```. 
For exemple :
```{r exemple1}
parameters$compaVD=c("Ctrast1-Ctrast2-Ctrast3") # this create 1 venn diagram
parameters$compaVD=c("Ctrast1-Ctrast2-Ctrast3", # this create 3 venn diagrams
                     "Ctrast4-Ctrast5-Ctrast6",
                     "Ctrast7-Ctrast8-Ctrast9")
```
Be careful, with the VD="both" you will only have to provide two contrasts.
Example:
```{r exemple2}
parameters$compaVD=c("Ctrast1-Ctrast2") # this create 1 venn diagram
parameters$compaVD=c("Ctrast1-Ctrast2", # this create 3 venn diagrams
                     "Ctrast1-Ctrast3",
                     "Ctrast2-Ctrast3") 
```
With our data, we will make 3 Venn diagrams for the different types (all, up and down).
```{r venn1, eval=FALSE}
parameters$compaVD = c("AC1vsAC2-AC1vsAC3-AC2vsAC3",
                       "BC1vsBC2-BC1vsBC3-BC2vsBC3",
                       "AC1vsBC1-AC2vsBC2-AC3vsBC3")

# graph type "all"
parameters$VD = "all"
VD(resDEG, parameters, asko_data)

# graph type "up"
parameters$VD = "up"
VD(resDEG, parameters, asko_data)

# graph type "down"
parameters$VD = "down"
VD(resDEG, parameters, asko_data)
```
\newpage To use the VD="both" option, we need to provide only two contrasts at a time.
```{r venn2, eval=FALSE}
# graph type "both"
parameters$compaVD = c("AC1vsBC1-AC2vsBC2",
                       "AC1vsBC1-AC3vsBC3",
                       "AC2vsBC2-AC3vsBC3")
parameters$VD = "both"
VD(resDEG, parameters, asko_data)
```
All graphs will appear in a folder named **"DEG_test/vennDiagram/"**. Some example of venn diagrams :
```{r venndiagram, echo=FALSE, out.width='48%', out.height='48%', fig.cap="Venn Diagrams", fig.align='center', fig.show='hold', fig.keep='all'}
par(mar=c(0,0,0,0), oma=c(0,0,0,0))
knitr::include_graphics(c("DEG_test/vennDiagram/AC1vsBC1-AC2vsBC2-AC3vsBC3_all.png","DEG_test/vennDiagram/AC1vsBC1-AC2vsBC2-AC3vsBC3_up.png","DEG_test/vennDiagram/AC1vsBC1-AC2vsBC2-AC3vsBC3_down.png","DEG_test/vennDiagram/AC1vsBC1-AC2vsBC2_mixed.png"))
```
\newpage

### Upset graphs
You can display all contrast, you just need to specify the type of comparison wanted ```parameters$upset_basic```:
 - "all"   : Create chart for all differentially expressed genes 
 - "up"    : Create chart for gene expressed UP
 - "down"  : Create chart for gene expressed DOWN
 - "mixed" : Create chart for gene expressed UP and DOWN (in the same graph) 
 - NULL    : Don't make graphs
You can display multiples graphs based on list of contrast ```parameters$upset_list```, you need to precise the type of comparison ```parameters$upset_type```. Example:
```{r exemple3}
# Precise type of comparison: all, down, up, mixed.
parameters$upset_type = "all" 
# Give a list of contrast, for example:
parameters$upset_list = c("Ctrast1-Ctrast2-Ctrast3")   # this create 1 graphs
parameters$upset_list = c("Ctrast1-Ctrast2-Ctrast3",   # this create 3 graphs
                          "Ctrast4-Ctrast5-Ctrast6",
                          "Ctrast1-Ctrast2-Ctrast3-Ctrast4-Ctrast5-Ctrast6")
```
With our data, we will make several upset charts for the different types (all, up, down and mixed), with all contrast and list of contrast.
```{r upset, eval=FALSE}
parameters$upset_list = c("AC1vsAC2-AC1vsAC3-AC2vsAC3",
                          "BC1vsBC2-BC1vsBC3-BC2vsBC3",
                          "AC1vsBC1-AC2vsBC2-AC3vsBC3")

# graphs type "all"
parameters$upset_basic = "all" # all contrast
parameters$upset_type = "all"  # list of contrast
UpSetGraph(resDEG, data, parameters)

# graphs type "mixed"
parameters$upset_basic = "mixed" # all contrast
parameters$upset_type = "mixed"  # list of contrast
UpSetGraph(resDEG, data, parameters)

# graphs type "up"
parameters$upset_basic = "up" # all contrast
parameters$upset_type = "up"  # list of contrast
UpSetGraph(resDEG, data, parameters)

# graphs type "down"
parameters$upset_basic = "down" # all contrast
parameters$upset_type = "down"  # list of contrast
UpSetGraph(resDEG, data, parameters)
```
An **"DEG_test/UpSetR_graphs/"** directory will be created with two subdirectories **"DEG_test/UpSetR_graphs/Global_upset/"** and **"DEG_test/UpSetR_graphs/ Subset_upset/"**.\
\newpage Some example of upset graphs (from subset "AC1vsBC1-AC2vsBC2-AC3vsBC3"):
```{r upsetR, echo=FALSE, out.width='48%', out.height='48%', fig.cap="Venn Diagrams", fig.align='center', fig.show='hold', fig.keep='all'}
par(mar=c(0,0,0,0), oma=c(0,0,0,0))
knitr::include_graphics(c("DEG_test/UpSetR_graphs/Subset_upset/DEG_test_UpSetR_AC1vsBC1-AC2vsBC2-AC3vsBC3_allDEG.png","DEG_test/UpSetR_graphs/Subset_upset/DEG_test_UpSetR_AC1vsBC1-AC2vsBC2-AC3vsBC3_upDEG.png","DEG_test/UpSetR_graphs/Subset_upset/DEG_test_UpSetR_AC1vsBC1-AC2vsBC2-AC3vsBC3_downDEG.png","DEG_test/UpSetR_graphs/Subset_upset/DEG_test_UpSetR_AC1vsBC1-AC2vsBC2-AC3vsBC3_mixedDEG.png"))
```
\newpage 

## GO Enrichment Analysis
We uses the GOs annotations file to perform enrichment analysis on gene differentially expressed. For this, you define :
 - ```parameters$GO_threshold``` &nbsp;the significant threshold used to filter p-values 
 - ```parameters$GO_max_top_terms``` &nbsp;the maximum number of GO terms plot 
 - ```parameters$GO_min_num_genes``` &nbsp;the minimum number of genes for each GO terms 
 - ```parameters$GO``` &nbsp;gene set chosen for analysis 'up', 'down', 'both' (up+down) 
 - ```parameters$GO_algo``` &nbsp;algorithms for runTest function ("classic", "elim", "weight", "weight01", "lea", "parentchild")
 - ```parameters$GO_stats``` &nbsp;statistical tests for runTest function ("fisher", "ks", "t", "globaltest", "sum", "ks.ties")
 - ```parameters$Ratio_threshold``` &nbsp;the min ratio for display GO in graph

After that, we can run Go enrichment analysis:
```{r GO1}
# Parameters
parameters$GO_threshold = 0.05 
parameters$GO_max_top_terms = 10
parameters$GO_min_num_genes = 10 
parameters$GO = "both" 
parameters$GO_algo = "weight01"
parameters$GO_stats = "fisher"
parameters$Ratio_threshold = 1
```
```{r GO2, eval=FALSE}
# run analysis
GOenrichment(resDEG, parameters)
```
An **"DEG_test/GO_images/"** directory will be created with all GO images and tables of statistics.\
Example of graph:
```{r GOgraph, echo=FALSE, out.width='48%', out.height='48%', fig.cap="Venn Diagrams", fig.align='center', fig.show='hold', fig.keep='all'}
par(mar=c(0,0,0,0), oma=c(0,0,0,0))
knitr::include_graphics(c("DEG_test/GO_images/AC1vsBC1_Pvalue_BUBBLESgraph.png","DEG_test/GO_images/AC1vsBC1_Ratio_BUBBLESgraph.png"))
```
\newpage Example of one staistical table:
\scriptsize

 GO.ID      | Term                | Annotated | Significant | Expected | statisticTest | Ratio        | GO_cat |
 -----------|---------------------|-----------|-------------|----------|---------------|--------------|--------|
 GO:0003735 | structural const... |       135 |          36 |    10.81 |       4.6e-11 | 3.3302497687 | MF     |
 GO:0000155 | phosphorelay sen... |        22 |           7 |     1.76 |        0.0012 | 3.9772727272 | MF     |
 GO:0003729 | mRNA binding        |        13 |           5 |     1.04 |        0.0024 | 4.8076923076 | MF     |
 GO:0036094 | small molecule b... |      1327 |         105 |   106.24 |        0.0038 | 0.9883283132 | MF     |
 ... |

\normalsize
Explications of some columns:

 - **Annotated:** number of genes in your genome are annotated with the GO-term.
 - **Significant:** number of genes belonging to your input which are annotated with the GO-term.
 - **Expected:** show an estimate of the number of genes a node of size Annotated would have if the significant genes were to be randomly selected from the gene universe. 
 - **statisticTest:** result of fisher test

\newpage

## Clustering and GO enrichment

A compléter par Kevin !

\newpage

# PARAMETERS AND COMMAND LINE REFERENCE
## files and samples options
\scriptsize

 command line | parameters | default | descriptions |
 ---------------|-------------------|---------|----------------------------------|
 ```-o``` or ```--out``` | ```parameters$analysis_name``` | DE_analysis | Output directory name (do not put space!) |
 ```-d``` or ```--dir``` | ```parameters$dir_path``` | ```"."``` | Work directory path |
 ```-O``` or ```--org``` | ```parameters$organism``` | Asko | Output files prefix |
 ```-f``` or ```--fileofcount``` | ```parameters$fileofcount``` | NULL | Matrix of count for all samples/conditions |
 ```-G``` or ```--col_genes``` | ```parameters$col_genes``` | 1 | Column of genes ids in count files |
 ```-C``` or ```--col_counts``` | ```parameters$col_counts``` | 7 | Column of counts in count files |
 ```-t``` or ```--sep``` | ```parameters$sep``` | NULL | Field separator for count files or count matrix |
 ```-c``` or ```--contrasts``` | ```parameters$contrast_file``` | NULL | Matrix of different contrasts desired |
 ```-s``` or ```--sample``` | ```parameters$sample_file``` | NULL | File describing the samples |
 ```-a``` or ```--annotation``` | ```parameters$annotation``` | NULL | File containing the genes' annotations |
 ```--ID2GO``` | ```parameters$geneID2GO_file``` | NULL | GO annotation files |
 ```-S``` or ```--select``` | ```parameters$select_sample``` | NULL | Selected sampls |
 ```-r``` or ```--remove``` | ```parameters$rm_sample``` | FALSE | Removed samples |
 ```-R``` or ```--regex``` | ```parameters$regex``` | FALSE | Use regex when selecting/removing samples |
 ```-k``` or ```--mk_context``` | ```parameters$mk_context``` | FALSE | Generate automatically the context names |

\normalsize
## Filter and normalization
\scriptsize

 command line | parameters | default | descriptions
 ----------------|-----------------------|-------|-----------------------------------------------|
 ```--th_cpm``` | ```parameters$threshold_cpm``` | 0.5 | CPM's threshold 
 ```--rep``` | ```parameters$replicate_cpm``` | 3 | Minimum of samples pass CPM's threshold 
 ```--norm_factor``` | ```parameters$norm_factor``` | FALSE | Generate file with normalize factor values 
 ```--norm_counts``` | ```parameters$norm_counts``` | FALSE | Generate files with mormalized counts 
 ```--dens_bottom_mar``` | ```parameters$densbotmar``` | 20 | Set bottom margin of density plot to help position the legend 
 ```--dens_inset``` | ```parameters$densinset``` | 0.45 | Set position the legend in bottom density graphe 
 ```--legend_col``` | ```parameters$legendcol``` | 6 | Set numbers of column for density plot legends 
 ```--palette``` | ```parameters$palette``` | Set2 | color palette (ggplot)
 ```--hm``` | ```parameters$heatmap``` | TRUE | Generation of the expression heatmap
 ```--nh``` | ```parameters$numhigh``` | 50 | Number of genes in the heatmap 
 ```--dclust``` | ```parameters$distcluts``` | euclidean | The distance measure to be used : euclidean, maximum, manhattan, canberra, binary or minkowski 
 ```--hclust``` | ```parameters$hclust``` | complete | The agglomeration method to be used : ward.D, ward.D2, single, complete, average, mcquitty, median or centroid

\normalsize
## Differential expression analysis
\scriptsize

 command line | parameters | default | descriptions
 ------------------|---------------------|-----|--------------------------------------|
 ```-n``` or ```--normalization``` | ```parameters$normal_method``` | TMN | normalization method (TMM/RLE/ upperquartile/none)
 ```--adj``` | ```parameters$p_adj_method``` | fdr | p-value adjust method (holm/hochberg/hommel/ bonferroni/BH/BY/fdr/none) 
 ```--th_FDR``` | ```parameters$threshold_FDR``` | 0.05 | FDR threshold
 ```--glm``` | ```parameters$glm``` | qlf | GLM method (lrt/qlf)
 ```--glmDisp``` | ```parameters$glm_disp``` | FALSE | Estimate Common, Trended and Tagwise Negative Binomial dispersions GLMs 
 ```--lfc``` | ```parameters$logFC``` | TRUE | logFC in the summary table 
 ```--th_lfc``` | ```parameters$threshold_logFC``` | 1 | logFC threshold 
 ```--fc``` | ```parameters$FC``` | TRUE | FC in the summary table 
 ```--lcpm``` | ```parameters$logCPM``` | FALSE | logCPm in the summary table 
 ```--fdr``` | ```parameters$FDR``` | TRUE | FDR in the summary table 
 ```--lr``` | ```parameters$LR``` | FALSE | LR in the summary table 
 ```--sign``` | ```parameters$Sign``` | TRUE | Significance (1/0/-1) in the summary table
 ```--expr``` | ```parameters$Expression``` | TRUE | Significance expression in the summary table 
 ```--mc``` | ```parameters$mean_counts``` | TRUE | Mean counts in the summary table  
 ```--plotMD``` | ```parameters$plotMD ```| FALSE | Mean-Difference Plot (aka MA plot) 
 ```--plotVO``` | ```parameters$plotVO``` | FALSE | Volcano plot 
 ```--glimMD``` | ```parameters$glimMD``` | FALSE | Glimma - Interactif Mean-Difference Plot (aka MA plot) 
 ```--glimVO``` | ```parameters$glimVO``` | FALSE | Glimma - Interactif Volcano plot 

\normalsize
## Venn and Upset graphs
\scriptsize

 command line | parameters | default | descriptions
 --------|--------------|---|---------------------------------------|
 ```--VD``` | ```parameters$VD``` | NULL | Plot VennDiagram, precise type of comparison: all, down, up or both 
 ```--compaVD``` | ```parameters$compaVD``` | NULL | Contrast comparison list to display in VennDiagram 
 ```--upset_basic``` | ```parameters$upset_basic``` | NULL | Display UpSetR charts for all contrasts, precise type of comparison: all, down, up, mixed.
 ```--upset_type``` | ```parameters$upset_type``` | NULL | Display UpSetR charts for list of contrasts, precise type of comparison: all, down, up, mixed. 
 ```--upset_list``` | ```parameters$upset_list``` | NULL | Contrast comparison list to display in UpSetR chart 

\normalsize
## GOs enrichment analysis
\scriptsize

 command line | parameters | default | descriptions
 --------------|---------------------|----|----------------------------------------|
 ```--GO``` | ```parameters$GO``` | NULL | GO enrichment analysis for gene expressed 'up', 'down', 'both', or NULL 
 ```--GO_algo``` | ```parameters$GO_algo``` | classic | algorithms which are accessible via the runTest function: "whichAlgorithms()"
 ```--GO_stats``` | ```parameters$GO_stats``` | fisher | statistical tests which are accessible via the runTest function: "whichTests()"
 ```--GO_cat``` | ```parameters$GO_cat``` | BP | GO category to explore: BP, CC or MF
 ```--GO_threshold``` | ```parameters$GO_threshold``` | 0.05 | the significant threshold used to filter p-values 
 ```--GO_max_top_terms``` | ```parameters$GO_max_top_terms``` | 10 | the maximum number of GO terms plot 
 ```--GO_min_num_genes``` | ```parameters$GO_min_num_genes``` | 10 | the minimum number of genes for each GO terms 
 ```--Ratio_threshold``` | ```parameters$Ratio_threshold``` | 0 | the minimum ratio for display GO in graph 

\normalsize
## Co-expression analysis
\scriptsize

 command line | parameters | default | descriptions
 ----------------------------|--------------------------------------|------|---------------------------------------|
 ```--coseq_model``` | ```parameters$coseq_model``` | Normal | Coseq model : Poisson, kmeans or Normal
 ```--coseq_normFactors``` | ```parameters$coseq_normFactors``` | none | Coseq normalization factor : TC , UQ, Med , DESeq, TMM or none (default none)
 ```--coseq_transformation``` | ```parameters$coseq_transformation``` | arcsin | Coseq tranformation : voom, logRPKM, arcsin, logit, logMedianRef, profile, logclr, clr, alr, ilr or none
 ```--coseq_ClustersNb``` | ```parameters$coseq_ClustersNb``` | 2:12 | Coseq : number of clusters desired (2:12 (auto), number from 2 to 12)
 ```--coseq_ContrastsThreshold``` | ```parameters$coseq_ContrastsThreshold``` | 1 | Coseq : number of contrasts in which DE genes are found for clustering 
 ```--coseq_normFactors``` | ```parameters$coseq_normFactors``` | none | Coseq normalization factor (TC , UQ, Med , DESeq, TMM , none)

\normalsize

